library/ecp_curves.c

```diff
@@ -1195,6 +1240,76 @@ static int ecp_mod_p255( mbedtls_mpi *N )
 }
 #endif /* MBEDTLS_ECP_DP_CURVE25519_ENABLED */
 
+#if defined(MBEDTLS_ECP_DP_CURVE448_ENABLED)
+
+/* Size of p448 in terms of mbedtls_mpi_uint */
+#define P448_WIDTH      ( 448 / 8 / sizeof( mbedtls_mpi_uint ) )
+
+/* Number of limbs fully occupied by 2^224 (max), and limbs used by it (min) */
+#define DIV_ROUND_UP( X, Y ) ( ( X + Y - 1 ) / ( Y ) )
+#define P224_WIDTH_MIN   ( 28 / sizeof( mbedtls_mpi_uint ) )
+#define P224_WIDTH_MAX   DIV_ROUND_UP( 28, sizeof( mbedtls_mpi_uint ) )
+#define P224_UNUSED_BITS ( ( P224_WIDTH_MAX * sizeof( mbedtls_mpi_uint ) * 8 ) - 224 )
+
+/*
+ * Fast quasi-reduction modulo p448 = 2^448 - 2^224 - 1
+ * Write N as A0 + 2^448 A1 and A1 as B0 + 2^224 B1, and return
+ * A0 + A1 + B1 + (B0 + B1) * 2^224.  This is different to the reference
+ * implementation of Curve448, which uses its own special 56-bit limbs rather
+ * than a generic bignum library.  We could squeeze some extra speed out on
+ * 32-bit machines by splitting N up into 32-bit limbs and doing the
+ * arithmetic using the limbs directly as we do for the NIST primes above,
+ * but for 64-bit targets it should use half the number of operations if we do
+ * the reduction with 224-bit limbs, since mpi_add_mpi will then use 64-bit adds.
+ */
+static int ecp_mod_p448( mbedtls_mpi *N )
+{
+    int ret;
+    size_t i;
+    mbedtls_mpi M, Q;
+    mbedtls_mpi_uint Mp[P448_WIDTH + 1], Qp[P448_WIDTH];
+
+    if( N->n <= P448_WIDTH )
+        return( 0 );
+
```

Comment:
@mpg @NWilson I think this and the analogous function for Curve25519 should be written in-place, for multiple reasons: (1) Performance: It's both faster and consumes less memory. (2) Security: I feel uneasy about using MPI's with stack-allocated limb-buffers, because some MPI API functions like `mpi_add_mpi` (which is used below for `M`!) assume that the limb-buffers have been obtained from `malloc`, and it is easy to run into a heap corruption using these functions with hand-crafted MPI's living on the stack.
@mpg @NWilson WDYT?

Reply:
You're right, it's playing with fire! But adding an allocation on a fast path isn't ideal (the modular reduction is done inside the exponentiation loop, for example).

I'd be happy to see this made safer - but equally, it's currently cunning enough not to cause the MPIs to be resized onto the heap, so it does work... just about.

Safety's probably more important, I agree. OpenSSL has an idea of a "context" where an operation like exponentiation can cache buffers, to avoid having to reallocate them inside the inner loop - something like that might be useful, for callers that invoke ecp_mod_p448 repeatedly?